Added migration page overview (#1061)
json-schema-org/website: pages/specification/migration/_index.md

Reviewer: I think you are missing covering the `maxDecimal` keyword

Author: Thank you for pointing them out.

Author: @jviotti The only detail I have on maxDecimal is that it was introduced in Draft 01. 
Could you point me to any data that would help?

cc; @jdesrosiers

Reviewer 2: `maxDecimal` was replaced by `divisibleBy` in draft-02.

section: docs
---

### Introduction

> Migrations are like version control for your system, allowing your team to define and share the schema progressions. If you have ever had to figure out when a keyword was introduced and how it behaved, you've faced the problem that migration guides solve.

Reviewer: I'm not sure I 100% follow the version control analogy (and not every JSON Schema user might necessarily understand version control). What about something more direct, like (total brainstorm, feel free to ignore 😅):

> Just as your schemas evolve over time, the JSON Schema specification does as well. This guide provides useful information to make upgrading your schemas across versions of the specification a breeze.

Or something like that?

Reviewer 2: This sounds like it's conflating the [schema migration pattern](https://en.wikipedia.org/wiki/Schema_migration) where database changes are stored in version control as scripts that modify the database schema. This is not related to that concept and has nothing to do with version control, even as an analogy.

Author: Thank you Juan.

> If you're moving from one Draft to another, learn how to use our migration guides and tooling ([alterschema](https://alterschema.sourcemeta.com/)) to bring your work with you.

Reviewer: ```suggestion
If you're moving from one Draft to another, learn how to use our migration guides and tooling ([AlterSchema](https://alterschema.sourcemeta.com/)) to bring your work with you.
```

### Keywords Overview

Here is a comprehensive overview to get you going.

> | All Keywords          | Specification          | Draft introduction | Removed | Changed                       |

Reviewer: I think there are some issues with the Specification column. We only have "Core" and "Validation" specifications, and those specifications may define multiple vocabularies (like "Applicator" and "Content"), which in turn define keywords of certain types like "Assertion" or "Annotation".

If the intention is to only highlight the specification (which I think it makes sense), then let's stick to only "Core" and "Validation"?

Reviewer: I think you might be conflating specification with vocabulary. For example, `contentSchema` is a keyword from the "Content" vocabulary, which is defined in the "Validation" spec

Author: Yes, I traced it to the [Specifications](https://json-schema.org/draft/2020-12/json-schema-core#section-10.3.1.1). Thank you.

| --------------------- | ---------------------- | ------------------ | ------- | ----------------------------- |
| $anchor               | Core                   | 2019-09            | No      | No (updated in 2020)          |

> | $comment              | Core                   | 07                 | Yes     | -                             |

Reviewer: Can we put all mentions of keywords in backticks? i.e. 

```suggestion
| `$comment`              | Core                   | 07                 | Yes     | -                             |
```

| $id                   | Core                   | 06                 | No      | Updated from id               |

> | $defs                 | Core                   | 2019-09            | No      | Updated from definitions      |

Reviewer: All of these keywords that were "updated from" other keywords should probably be marked as removed? i.e. `definitions` is supposed to be gone after the introduction of `$defs`

Reviewer 2: Maybe something like "replaces" and "replaced by" instead of "updated to/from".

| pattern               | Core                   | 03                 | No      | No                            |
| patternProperties     | Validation             | 03                 | No      | No                            |
| prefixItems           | Applicator             | 2020-12            | No      | Updated from items            |
| propertyNames         | Validation             | 06                 | No      | No                            |
| readOnly              | Validation             | 07                 | No      | No                            |
| required              | Validation             | 03                 | No      | No                            |
| requires              | Validation             | 02                 | Yes     | -                             |
| title                 | Annotation             | 01                 | No      | No                            |
| then                  | Validation             | 07                 | No      | No                            |
| type                  | Validation             | 02                 | No      | No                            |

> | unevaluatedItems      | Applicator, Annotation | 2019-09            | No      | No                            |

Reviewer: There is no "Annotation" specification. Only "Core" and "Validation"

Author: Thank you, I will effect the changes.

| unevaluatedProperties | Applicator, Annotation | 2019-09            | No      | No                            |

> | uniqueItems           | Assertion              | 02                 | Yes     | No                            |

Reviewer: There is no "Assertion" specification. Only "Core" and "Validation".

| writeOnly             | Validation             | 07                 | No      | No                            |

> For a detailed read-through about all the changes see each Draft migration guide.

Reviewer: Maybe link to them here?

Author: Yes, they are yet to be available. I will link them once they're available.

section: docs
---

### Introduction

> Just as your schemas evolve over time, the JSON Schema specification does as well. This guide provides useful information to make upgrading your schemas across versions of the specification a breeze.

Reviewer 3: comment: avoid using subjective terms like "breeze". While it might be intended to reassure the reader, it could also set unrealistic expectations. What one person considers easy, another might find challenging. Using such a subjective term can lead to frustration for readers who struggle with the task.

suggestion: "This guide provides useful information to help you upgrade your schemas across versions of the specification."

Author: Noted and applied.

> Learn how to use our migration guides and tooling ([AlterSchema](https://alterschema.sourcemeta.com/)) to bring your work with you.

Reviewer: I feel "to bring your work with you" is weird here? Maybe we want to say that these things will help you or something like that?

### Keywords Overview

> Here is a comprehensive overview to get you going.

Reviewer 4: My version - Here is a comprehensive overview to get you **started**

I read a pointer on Google style guide to draft this way, but I'm unable to recall the exact one. Please ignore if not relevant.

Author: Agreed Anitha, I will use your version. Thank you.

> Here is a comprehensive overview to get you started.

Reviewer 3: comment: I find this sentence incomplete as an introduction to the table. What type of info does the table contain? 

suggestion: "This table provides an overview of keyword changes and removals in JSON Schema specifications from 2019-09 to the latest draft."

Author: Noted and applied.

> | All Keywords                | Specification | Draft introduction | Removed | Changed                                                                                      |

Reviewer 3: question: @jdesrosiers and @jviotti, based on its contents, shouldn't the "Specification" column be named "Vocabulary"?

Reviewer 2: The content is consistent with its heading. "Core", "Validation", and "Hyper-Schema" are the names of the two main spec documents. It is confusing because those are also names of vocabularies.

| --------------------------- | ------------- | ------------------ | ------- | -------------------------------------------------------------------------------------------- |

> | `$anchor`                   | Core          | 2019-09            | No      | Partially replaced `$id`                                                                     |

Reviewer: I think we should improve on the description of the change. Instead of just saying it partially replaces `$id`, we can mention something like the fact that `$id`s that defined anchor-only URI, like `$id: #foo` are now replaced with `$anchor`?

Reviewer 2: My assumption is that this needs to be very brief to not explode the table. I'm not sure we have the space to explain the change properly. My understanding was that this kind of thing was going to be covered in detail elsewhere.

| `$comment`                  | Core          | 07                 | No      |                                                                                              |
| `$id`                       | Core          | 06                 | No      | Replaced `id`                                                                                |
| `$defs`                     | Core          | 2019-09            | No      | Replaced `definitions`                                                                       |
| `$dynamicAnchor`            | Core          | 2020-12            | No      | Replaced `$recursiveAnchor`                                                                  |

> | `$dynamicRef`               | Core          | 2020-12            | No      | Replaced `$recursiveRef`                                                                     |

Reviewer 5: Hi, I think because `$dynamicRef`, `$dynamicAnchor` added more flexibility / usage over $`recursiveAnchor`.
Maybe we can say it as replacement or enhancement ? OR something like that

| `$recursiveAnchor`          | Core          | 2019-09            | 2020-12 | Replaced by `$dynamicAnchor`                                                                 |
| `$recursiveRef`             | Core          | 2019-09            | 2020-12 | Replaced by `$dynamicRef`                                                                    |

> | `$ref` (legacy)             | Core          | 03                 | 2019-09 | Replaced by `$ref`                                                                           |

Reviewer: This entry feels weird. We are saying that `$ref` was removed on `2019-09` and replaced by `$ref`, which feels very confusing. The change this specific entry should be documenting is that before 2019-09, `$ref` took precedence over any other sibling keyword. In other words, before 2019-09, if you put any other keywords alongside `$ref`, it would be ignored.

> | `$ref`                      | Core          | 2019-09            | No      | Replaced `$ref` (legacy)                                                                     |

Reviewer 2: If describing the change in more detail is an option, we wouldn't need two entries.

| `$schema`               | Core          | 03                 | No      | No                             |
| `$vocabulary`           | Core          | 2019-09            | No      | No                             |
| `additionalItems`       | Core          | 03                 | 2020-12 | Replaced by `items`            |
| `allOf`                 | Core          | 04                 | No      | No                             |
| `anyOf`                 | Core          | 04                 | No      | No                             |
| `const`                 | Validation    | 06                 | No      | No                             |
| `contains`              | Core          | 06                 | No      | No                             |
| `contentEncoding` (legacy) | Core          | 01                 | 04      | Replaced by `media`.`binaryEncoding`                 |

> | `media`                    | Validation    | 04                 | 07      | Replaced `conentEncoding` (legacy) and replaced by `contentEncoding` and `contentMediaType` |

Reviewer 2: I missed this.

```suggestion
| `media`                    | Hyper-Schema  | 04                 | 07      | Replaced `conentEncoding` (legacy) and replaced by `contentEncoding` and `contentMediaType` |
```

> | `media`                    | Validation    | 04                 | 07      | Replaced `conentEncoding` (legacy) and replaced by `contentEncoding` and `contentMediaType` |

Reviewer 4: Replaced **`conentEncoding`** (legacy) and replaced by `contentEncoding` and `contentMediaType` |

Typo - contentEncoding

> | `additionalItems`           | Core          | 03                 | 2020-12 | Replaced by `items`                                                                          |

Reviewer: In 2019-09, this keyword was modified to produce annotations

Reviewer 2: I don't think we need to note that things started producing annotations in 2019-09. Nothing produced annotations before 2019-09. I think it's given or at least can be generally stated once at the top of the table that annotations only became a thing in 2019-09.

> | `contains`                  | Core          | 06                 | No      | No                                                                                           |

Reviewer: The `contains` keyword was slightly changed in 2020-12. In 2020-12, it now emits annotations and affects `unevaluatedItems`

> | `contentEncoding` (legacy)  | Core          | 01                 | 04      | Replaced by `media`.`binaryEncoding`                                                         |

Reviewer: I don't think these `contentEncoding` entries make sense. `contentEncoding` as a keyword remains even up to 2020-12

> | `media`                     | Hyper-Schema  | 04                 | 07      | Replaced `contentEncoding` (legacy) and replaced by `contentEncoding` and `contentMediaType` |

Reviewer: The keywords in this table are order alphabetically, so `media` here doesn't make sense?

Reviewer 2: I only suggested including `media` in order to tell the complicated story of `contentEncoding`. In the beginning, there was only one spec for core, validation, and hyper-schema. I think it would make sense if we considered the original `contentEncoding` part of hyper-schema even though it wasn't a separate spec yet. Then we can remove these two entries and pretend it was introduced in draft-07 even though it's a little more complicated than that.

> | `contentEncoding`           | Validation    | 07                 | No      | Replaced `media`.`binaryEncoding`                                                            |

Reviewer: I don't think the notes make sense. `contentEncoding` never replaced any other keyword. That said, on 2019-09, it became an annotation keyword

> | `contentMediaType`          | Validation    | 07                 | No      | Replaced `media`.`type`                                                                      |

Reviewer: Same here. I don't think we should bother with Hyper-Schema, and this keyword also became an annotation in 2019-09

Author: By understanding, the specification covers - Core and Validation. 
The hyperschema is a hypermedia vocabulary—I don't think we should have it here; otherwise, we may have to include other hyperschema keywords not considered for dialect migration.

cc: @jdesrosiers

Author: @jdesrosiers I am still waiting to hear from you. Thank you

Reviewer 2: Sorry for the slow response. I've been away almost all of last week and I'm catching up now.

> The hyperschema is a hypermedia vocabulary—I don't think we should have it here

There are three specifications: Core, Validation, and Hyper-Schema. Each spec defines one or more vocabularies.

- Core: core, applicators, unevaluated
- Validation: validation, format-annotation, format-assertion, meta, content
- Hyper-Schema: hyper-schema

So, Hyper-Schema is a specification that defines the hyper-schema vocabulary. I was referring to "Hyper-Schema" the spec.

> otherwise, we may have to include other hyperschema keywords not considered for dialect migration.

I addressed that in a [previous comment](https://github.com/json-schema-org/website/pull/1061#discussion_r1825204529). I suggested including it because it was previously moved from Hyper-Schema to Validation and it didn't make sense to say it replaced the old Hyper-Schema keyword without also listing the keyword it replaced. The comment I linked suggests an alternative.

| `default`                   | Validation    | 01                 | No      | No                                                                                           |
| `dependencies`              | Validation    | 03                 | 2019-09 | Replaced `requires` and replaced by `dependentSchemas` and `dependentRequired`               |
| `dependentRequired`         | Validation    | 2019-09            | No      | Partially replaced `dependencies`                                                            |
| `dependentSchemas`          | Core          | 2019-09            | No      | Partially replaced `dependencies`                                                            |
| `deprecated`                | Validation    | 2019-09            | No      | No                                                                                           |
| `disallow`                  | Core          | 01                 | 04      | Replaced by `not`                                                                            |

> | `divisibleBy`               | Validation    | 02                 | No      | Replaced by `multipleOf`                                                                     |

Reviewer: This keyword was removed when it was replaced by `multipleOf`

Reviewer 2: It was removed in draft-04

Author: Noted and adjusted.

| `else`                      | Core          | 07                 | No      | No                                                                                           |
| `enum`                      | Validation    | 01                 | No      | No                                                                                           |

> | `examples`                  | Validation    | 06                 | No      | No                                                                                           |

Reviewer: In 2019-09, this keyword now produces an annotation

| `exclusiveMaximum` (legacy) | Validation    | 03                 | 06      | Replaced `maximumCanEqual`                                                                   |
| `exclusiveMaximum`          | Validation    | 06                 | No      | Replaced `exclusiveMaximum` (legacy)                                                         |
| `exclusiveMinimum` (legacy) | Validation    | 03                 | 06      | Replaced `minimumCanEqual`                                                                   |
| `exclusiveMinimum`          | Validation    | 06                 | No      | Replaced `exclusiveMinimum` (legacy)                                                         |
| `extends`                   | Core          | 01                 | 04      | Replaced by `allOf`                                                                          |

> | `format`                    | Validation    | 01                 | No      | -                                                                                            |

Reviewer: This keyword was turned into an annotation keyword in 2019-09. Also there were many changes in the set of permitted values this keyword could take over varios versions of the specifications. Not sure if we want to document those here

| `if`                    | Core          | 07                 | No      | No                                                                             |
| `items`                 | Core          | 01                 | No      | Changed in 2020-12                                                            |
| `maxContains`           | Validation    | 2019-09            | No      | No                                                                             |
| `maxDecimals`           | Validation    | 01                 | Yes     | Replaced by `divisibleBy`                                                      |

> | `maxProperties`         | Validation    | 04                 | No      | No                                                                             |

Reviewer 2: I noticed `maxItems` and `minItems` are missing entirely in this PR. They've always existed since draft-01 and have never changed.

> | `items` (legacy)            | Core          | 01                 | 2020-12 | Replaced by `prefixItems`                                                                    |

Reviewer 4: | `items` (legacy)            | Core          | 01                 | 2020-12 | **Replaced by `items`**

> | `items` (legacy)            | Core          | 01                 | 2020-12 | Replaced by `prefixItems`                                                                    |

Reviewer 5: `items` Array  is replaced by `prefixItems`, and `items` Object is replaced by `additionalItems`...

> | `items` (legacy)            | Core          | 01                 | 2020-12 | Replaced by `prefixItems`                                                                    |

Reviewer 2: Technically, the object form of `items` was removed entirely and `additionalItems` was renamed `items`. So, it was partially replaced and partially removed. I don't know how detailed we need to be about that in this table. It's just an overview.

| `prefixItems`           | Core          | 2020-12            | No      | Replaced `items`               |
| `propertyNames`         | Core          | 06                 | No      | No                             |
| `readOnly`              | Validation    | 01                 | No      | No                             |
| `required`              | Validation    | 03                 | No      | No                             |
| `requires`              | Core          | 01                 | 03      | Replaced by `dependencies`     |
| `title`                 | Validation    | 01                 | No      | No                             |
| `then`                  | Core          | 07                 | No      | No                             |
| `type` (legacy)         | Core          | 01                 | 04      | Replaced by `type`             |
| `type`                  | Validation    | 04                 | No      | Replaced `type` (legacy)       |
| `unevaluatedItems`      | Core          | 2019-09            | No      | No                             |
| `unevaluatedProperties` | Core          | 2019-09            | No      | No                             |

> | `uniqueItems`           | Core          | 02                 | Yes     | No                             |

Reviewer: I think its even more complicated, as for keywords like this one, older specifications had it in the "Core" one but now they are on "Validation". Given that, maybe we should completely drop the "Specification" column?

Reviewer: Ah, I see you often do have multiple entries with different specifications, like with `type`

Author: @jviotti 
We are still going to have individual pages for the migration. This is meant to give an overview, and somehow, creating an individual page defeats the purpose of an overview. 

We can make it work as it is. We only need to get accurate information.

Reviewer 2: > older specifications had it in the "Core" one but now they are on "Validation"

What I was trying to do with this in my suggestions was to use the specification that it last appeared in. I may have messed up some of my suggestions or just missed some, but that's what I was trying to do. So, `uniqueItems` would be in `Validation`. If you have a better way to communicate this, please do that instead.

Reviewer 2: > We are still going to have individual pages for the migration.

If this table is just an overview and there's going to be somewhere else where some of these complicated situations appear, it wouldn't be too bad to oversimplify some of the more complicated situations here. Again, it's up to you.

Author: For the overview page, like in the case of `uniqueItems`, I will stick to the current specification it belongs to - `validation`. In Draft 2, we can add details of its change in the details section.

> | `optional`              | Core          | 02                 | Yes     | Replaced by `required`                                                                       |

Reviewer 2: ```suggestion
| `optional`              | Core          | 02                 | 03      | Replaced by `required`                                                                       |
```

> | `optional`                  | Core          | 02                 | No      | Replaced by `required`                                                                       |

Reviewer 5: @kwennB  `optional` is removed

Reviewer 2: It was removed in draft-03

Author: Noted and applied.

> | `patternProperties`         | Core          | 03                 | No      | No                                                                                           |

Reviewer: In 2019-09, this keyword started emitting annotations

> | `readOnly`                  | Validation    | 01                 | No      | No                                                                                           |

Reviewer: In 2019-09, this keyword started emitting annotations

> | `title`                     | Validation    | 01                 | No      | No                                                                                           |

Reviewer: In 2019-09, this keyword started emitting annotations

> | `type` (legacy)             | Core          | 01                 | 04      | Replaced by `type`                                                                           |

Reviewer: This is also weird. I think it's confusing to say `type` was replaced by `type`. I think the main change you want to document here is that previously, `type` could be set to a schema

Reviewer: The `any` type was previously valid too, and now it's not. Maybe worth adding a note

Author: @jviotti How best would you say to represent `type`?

Reviewer: To be honest, I'm not sure. Maybe this is OK, and we can just clarify that `type` no longer takes `any` and subschemas in its array form on the notes?

Reviewer 2: I think this depends on how much we can put in the "Changed" column. If a brief description of the change can fit in that column, we don't need two entries. The reason I suggested two entries was because I didn't think that was an option. So, I recommend either a description of the changes in "Changes" _or_ two entries, but not both.

> | `uniqueItems`               | Validation    | 02                 | Yes     | No                                                                                           |

Reviewer: This keyword was never removed, though it says "Yes" here

Author: Noted and applied.

> | `writeOnly`                 | Validation    | 07                 | No      | No                                                                                           |

Reviewer: In 2019-09, this keyword started emitting annotations

### Introduction

Just as your schemas evolve over time, the JSON Schema specification does as well. This guide provides useful information to make upgrading your schemas across versions of the specification a breeze.

Learn how to use our migration guides and tooling ([AlterSchema](https://alterschema.sourcemeta.com/)) to bring your work with you.

### Keywords Overview

Here is a comprehensive overview to get you going.

| All Keywords            | Specification | Draft introduction | Removed | Changed                        |
| ----------------------- | ------------- | ------------------ | ------- | ------------------------------ |

> | `$anchor`               | Core          | 2019-09            | No      | No (updated in 2020)           |

Author: Let's say "Partially replaced `$id`"

| `$ref`                  | Core          | 2019-09            | No      | No                             |
| `$schema`               | Core          | 02                 | No      | No                             |
| `$vocabualry`           | Core          | 2019-09            | No      | No                             |
| `additionalItems`       | Core          | 03                 | No      | No                             |
| `allOf`                 | Core          | 04                 | No      | No                             |
| `anyOf`                 | Core          | 04                 | No      | No                             |
| `const`                 | Validation    | 06                 | No      | No                             |
| `contains`              | Core          | 06                 | No      | No                             |
| `contentEncoding`       | Validation    | 07                 | No      | No                             |
| `contentMediaType`      | Validation    | 07                 | No      | No                             |
| `contentSchema`         | Validation    | 2019-09            | No      | No                             |

> | `definitions`           | Validation    | 04                 | No      | Replaced by` $defs`            |

Reviewer 2: ```suggestion
| `definitions`           | Validation    | 04                 | 2019-09 | Replaced by` $defs`            |
```

| `default`               | Validation    | 01                 | No      | No                             |
| `dependencies`          | Core          | 03                 | No      | No                             |
| `dependentRequired`     | Validation    | 2019-09            | No      | No                             |
| `dependentSchemas`      | Core          | 2019-09            | No      | Replaced `dependencies`        |
| `deprecated`            | Validation    | 2019-09            | No      | No                             |

> | `disallow`              | Validation    | 03                 | Yes     | No                             |

Reviewer 2: `disallow` was effectively the inverse of `type`. That was make unnecessary when `not` was introduced and could inverse any statement, not just `type`. So, I'm not sure saying it was replaced by `not` really makes sense by itself without more explanation.

```suggestion
| `disallow`              | Core          | 01                 | 04      | Replaced by `not`              |
```

| `divisibleBy`           | Validation    | 02                 | No      | Replaced by `multipleOf`       |
| `else`                  | Core          | 07                 | No      | No                             |
| `enum`                  | Validation    | 01                 | No      | No                             |
| `examples`              | Validation    | 06                 | No      | No                             |

> | `exclusiveMaximum`      | Validation    | 03                 | No      | Replaced `maximumCanEqual`     |
> | `exclusiveMinimum`      | Validation    | 03                 | No      | Replaced `minimumCanEqual`     |

Reviewer 2: This is another case where the keyword changed behavior without changing its name.

```suggestion
| `exclusiveMaximum` (legacy) | Validation    | 03                 | 06      | Replaced `maximumCanEqual`           |
| `exclusiveMaximum`          | Validation    | 06                 | No      | Replaced `exclusiveMaximum` (legacy) |
| `exclusiveMinimum` (legacy) | Validation    | 03                 | 06      | Replaced `minimumCanEqual`           |
| `exclusiveMinimum`          | Validation    | 06                 | No      | Replaced `exclusiveMinimum` (legacy) |
```

> | `extends`               | Validation    | 03                 | Yes     | No                             |

Reviewer 2: ```suggestion
| `extends`               | Core          | 01                 | 04      | Replaced by `allOf`            |
```

> | `format`                | Validation    | 02                 | No      | -                              |

Reviewer 2: ```suggestion
| `format`                | Validation    | 01                 | No      | -                              |
```

> | `id`                    | Core          | 04                 | No      | Replaced by `$id`              |

Reviewer 2: ```suggestion
| `id`                    | Core          | 03                 | 06      | Replaced by `$id`              |
```

| `if`                    | Core          | 07                 | No      | No                             |

> | `items`                 | Core          | 2019-09            | No      | Replaced by `prefixItems`      |

Reviewer 2: Yet another complicated story.

```suggestion
| `items` (legacy)        | Core          | 01                 | 2020-12 | Replaced by `prefixItems`                       |
| `items`                 | Core          | 2020-12            | No      | Replaced `additionalItems` and `items` (legacy) |
```

| `maxContains`           | Validation    | 2019-09            | No      | No                             |
| `maxProperties`         | Validation    | 04                 | No      | No                             |

> | `maximumCanEqual`       | Validation    | 02                 | No      | Replaced by `exclusiveMaximum` |
> | `minimumCanEqual`       | Validation    | 02                 | No      | Replaced by `exclusiveMinimum` |

Reviewer 2: ```suggestion
| `maximumCanEqual`       | Validation    | 01                 | 03      | Replaced by `exclusiveMaximum` |
| `minimumCanEqual`       | Validation    | 01                 | 03      | Replaced by `exclusiveMinimum` |
```

| `minContains`           | Validation    | 2019-09            | No      | No                             |
| `minProperties`         | Validation    | 04                 | No      | No                             |
| `multipleOf`            | Validation    | 04                 | No      | Replaced `divisibleBy`         |
| `not`                   | Core          | 04                 | No      | No                             |
| `oneOf`                 | Core          | 04                 | No      | No                             |

> | `optional`              | Validation    | 02                 | No      | Replaced by `required`         |

Reviewer 2: ```suggestion
| `optional`              | Core          | 02                 | No      | Replaced by `required`         |
```

> | `pattern`               | Core          | 03                 | No      | No                             |

Reviewer 2: ```suggestion
| `pattern`               | Core          | 01                 | No      | No                             |
```

| `patternProperties`     | Core          | 03                 | No      | No                             |
| `prefixItems`           | Core          | 2020-12            | No      | Replaced `items`               |
| `propertyNames`         | Core          | 06                 | No      | No                             |

> | `readOnly`              | Validation    | 07                 | No      | No                             |

Reviewer 2: ```suggestion
| `readOnly`              | Validation    | 01                 | No      | No                             |
```

| `required`              | Validation    | 03                 | No      | No                             |

> | `requires`              | Validation    | 02                 | Yes     | -                              |

Reviewer 2: ```suggestion
| `requires`              | Core          | 01                 | 03      | Replaced by `dependencies`     |
```

| `title`                 | Validation    | 01                 | No      | No                             |
| `then`                  | Core          | 07                 | No      | No                             |

> | `type`                  | Validation    | 02                 | No      | No                             |

Reviewer 2: `type` had a significant change in draft-04. It used to allow schemas as values and "any".

```suggestion
| `type` (legacy)         | Core          | 01                 | 04      | Replaced by `type`             |
| `type`                  | Validation    | 04                 | No      | Replaced `type` (legacy)       |
```

| `unevaluatedItems`      | Core          | 2019-09            | No      | No                             |
| `unevaluatedProperties` | Core          | 2019-09            | No      | No                             |

> | `uniqueItems`           | Assertion     | 02                 | Yes     | No                             |

Reviewer 2: ```suggestion
| `uniqueItems`           | Core          | 02                 | Yes     | No                             |
```
